### ISSUE TITLE
Function names can cause obligation solving to fail

### ISSUE THREAD

**Issue by johnli0135** [mattam82/Coq-Equations#273]
The following mutually recursive bundle (f, g, h) produces 3 obligations, which can all be discharged easily:
```coq
Inductive T := C : list T -> T.

Equations f (p : bool) (e : T) : T by struct e := {
  f false e := e ;
  f p (C xs) :=
    let xs := g true xs in
    C xs }
where g (p : bool) (xs : list T) : list T by struct xs := {
  g false xs := xs ;
  g _ [] := [];
  g p (x :: xs) :=
    let x := h true x in
    let xs := g true xs in
    x :: xs }
where h (p : bool) (x : T) : T by struct x := {
  h false x := x ;
  h p x :=
    let x := f true x in
    x }.
Next Obligation. destruct xs; auto. Defined.
Next Obligation. destruct x; auto. Defined.
Next Obligation. destruct x; auto. Defined.
```

If `h` is renamed to `c` however, discharging obligations one by one using Next Obligation fails:
```coq
Inductive T := C : list T -> T.

Equations f (p : bool) (e : T) : T by struct e := {
  f false e := e ;
  f p (C xs) :=
    let xs := g true xs in
    C xs }
where g (p : bool) (xs : list T) : list T by struct xs := {
  g false xs := xs ;
  g _ [] := [];
  g p (x :: xs) :=
    let x := c true x in
    let xs := g true xs in
    x :: xs }
where c (p : bool) (x : T) : T by struct x := {
  c false x := x ;
  c p x :=
    let x := f true x in
    x }.
(* c < g lexicographically ==>
   Next Obligation selects c obligations *)
Next Obligation. destruct x; auto. Defined.
Next Obligation. destruct x; auto. Defined.
(*
Solving obligations automatically...
Warning:
In environment
eos : Tactics.end_of_section
H : forall (p : bool) (e : T), f_graph p e (f p e)
g : forall (p : bool) (xs : list T), g_graph p xs (B.g p xs)
xs : list T
Unable to unify "B.g false xs" with "xs".
1 obligation remaining
Obligation 1 of f_graph_correct:
((forall (p : bool) (e : T), f_graph p e (f p e)) *
 ((forall (p : bool) (xs : list T), g_graph p xs (g p xs)) * (forall (p : bool) (x : T), c_graph p x (c p x))))%type.
*)
```

To get around this, one has to explicitly select g's obligation to solve first:
```coq
Obligation 1 of g_equation_3. destruct xs; auto. Defined.
Next Obligation. destruct x; auto. Defined.
Next Obligation. destruct x; auto. Defined.
```

**Comment by mattam82** [comment]
Hmm, there are cascading problem here. One thing is that Equations should figure out that [xs] for [g] and [x] for [h/c] have to be destructed to solve the equation obligations as they are recursive definitions. Proving them by hand in any order should have sufficed, however the logic for launching the graph correctness proof was buggy and started with an unsolved g_equation_3 here, resulting in failure to prove the theorem. The fix is in PR #277

**Comment by mattam82** [comment]
Fixed now.
